optuna/visualization/_timeline.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+import datetime
+from typing import NamedTuple
+
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study import Study
+from optuna.trial import TrialState
+from optuna.visualization._plotly_imports import _imports
+from optuna.visualization._utils import _make_hovertext
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+class _TimelineBarInfo(NamedTuple):
+    number: int
+    start: datetime.datetime
+    end: datetime.datetime
```

Comment:
It would be better to use `complete` instead of `end`.

Reply:
Agreed. I revised the code.